narwhal/primary/src/proposer.rs

```diff
@@ -581,16 +581,25 @@ impl Proposer {
                             let _ = self.tx_narwhal_round_updates.send(self.round);
                             self.last_parents = parents;
 
-                            // we re-calculate the timeout to give the opportunity to the node
-                            // to propose earlier if it's a leader for the round
-                            // Reschedule the timer.
+                            // Extend max_delay_timer to properly wait for leader from the
+                            // previous round.
+                            //
+                            // But min_delay_timer should not be extended: the network moves at
+                            // the interval of min_header_delay. Delaying header creation for
+                            // another min_header_delay after receiving parents from a higher
+                            // round and cancelling proposing, makes it very likely that higher
+                            // round parents will be received and header creation will be cancelled
+                            // again. So min_delay_timer is disabled to get the proposer in sync
+                            // with the quorum.
+                            // If the node becomes leader, disabling min_delay_timer to propose as
+                            // soon as possible is the right thing to do as well.
                             let timer_start = Instant::now();
                             max_delay_timer
                                 .as_mut()
                                 .reset(timer_start + self.max_delay());
                             min_delay_timer
                                 .as_mut()
-                                .reset(timer_start + self.min_delay());
```

Comment:
Back then I did perform a check on the left over deadline time https://github.com/MystenLabs/sui/pull/7782#discussion_r1092733743 before taking the decision to reset. Do we consider any benefit doing this, or just straight reset would be more desirable?

Reply:
I thought about resetting only for leader and keep the leftover deadline, but I think it is less predictable, and does not reduce delay enough when there are a lot of time left on the min_delay_timer.